C compiler error when default initializing an object field function when the default function is overloaded

### Description

I get a C compiler error when trying to compile this code with `nim c main.nim`:
```nim
proc evaluate*(params: int) =
    discard

proc evaluate*() =
    discard

type SearchInfo = object
  evaluation: proc() = evaluate

var a = SearchInfo()
```

```haskell
error: assignment to ‘void (*)(void *)’ from incompatible pointer type ‘tyProc__T4eqaYlFJYZUv9aG9b1TV0bQ’ {aka ‘void (*)(void)’}
```

### Nim Version

Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2024-04-27
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: f682dabf71f35833e934a79007906f94d8d32c92
active boot switches: -d:release


### Current Output

```text
Hint: used config file '/home/tsoj/.choosenim/toolchains/nim-#devel/config/nim.cfg' [Conf]
Hint: used config file '/home/tsoj/.choosenim/toolchains/nim-#devel/config/config.nims' [Conf]
.....................................................................
/tmp/main.nim(10, 5) Hint: 'a' is declared but not used [XDeclaredButNotUsed]
CC: main.nim
/home/tsoj/.cache/nim/main_d/@mmain.nim.c: In function ‘NimMainModule’:
/home/tsoj/.cache/nim/main_d/@mmain.nim.c:227:19: error: assignment to ‘void (*)(void *)’ from incompatible pointer type ‘tyProc__T4eqaYlFJYZUv9aG9b1TV0bQ’ {aka ‘void (*)(void)’} [-Wincompatible-pointer-types]
  227 |         T1_.ClP_0 = colontmpD_; T1_.ClE_0 = NIM_NIL;
      |                   ^
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread   -I'/home/tsoj/.choosenim/toolchains/nim-#devel/lib' -I/tmp -o /home/tsoj/.cache/nim/main_d/@mmain.nim.c.o /home/tsoj/.cache/nim/main_d/@mmain.nim.c' failed with exit code: 1
```


### Expected Output

```text
Should compile, or alternatively give a good error message
```


### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -257,6 +257,22 @@ proc isRecursiveType(t: PType, cycleDetector: var IntSet): bool =
   else:
     return false
 
+proc annotateClosureConv(n: PNode) =
+  case n.kind
+  of {nkNone..nkNilLit}:
+    discard
+  of nkTupleConstr:
+    if n.typ.kind == tyProc and n.typ.callConv == ccClosure and
+        n[0].typ.kind == tyProc and n[0].typ.callConv != ccClosure:
+      # restores `transf.generateThunk`
+      n[0] = newTreeIT(nkHiddenSubConv, n[0].info, n.typ,
+                       newNodeI(nkEmpty, n[0].info), n[0])
+      n.transitionSonsKind(nkClosure)
+      n.flags.incl nfTransf
+  else:
+    for i in 0..<n.len:
+      annotateClosureConv(n[i])
+
 proc fitDefaultNode(c: PContext, n: PNode): PType =
   inc c.inStaticContext
   let expectedType = if n[^2].kind != nkEmpty: semTypeNode(c, n[^2], nil) else: nil
@@ -272,6 +288,7 @@ proc fitDefaultNode(c: PContext, n: PNode): PType =
         # `changeType` to infer types for constant values
         # that's also the reason why we don't use `semExpr` to check
         # the type since two overlapping error messages might be produced
+    annotateClosureConv(n)
     result = n[^1].typ
   else:
     result = n[^1].typ

--- a/tests/objects/tobject_default_value.nim
+++ b/tests/objects/tobject_default_value.nim
@@ -776,5 +776,18 @@ template main {.dirty.} =
     var d: limited_int;
     doAssert d == 1
 
+  block: # bug #23545
+    proc evaluate(params: int) =
+        discard
+
+    proc evaluate() =
+        discard
+
+    type SearchInfo = object
+      evaluation: proc() = evaluate
+
+    var a = SearchInfo()
+    a.evaluation()
+
 static: main()
 main()

